File: python/BinaryTreePostorderTraversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def postorderTraversal(self, root):
        stack = []
        lastVisited = None
        node = root
        result = []
        while len(stack) > 0 or node:
            if node:
                stack.append(node)
                node = node.left
            else:
                curr = stack[-1]
                if not curr.right or lastVisited == curr.right:
                    out = stack.pop()
                    result.append(out.val)
                    lastVisited = out
                else:
                    node = curr.right
        return result

File: c++/BinaryTreePostorderTraversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode *root) {
        TreeNode *node = root;
        TreeNode *lastVisited;
        stack<TreeNode*> s;
        vector<int> result;
        while (!s.empty() || node) {
            if (node) {
                s.push(node);
                node = node->left;
            } else {
                TreeNode *peak = s.top();
                if (peak->right && lastVisited != peak->right) {
                    node = peak->right;
                } else {
                    result.push_back(peak->val);
                    s.pop();
                    lastVisited = peak;
                }
            }
        }
        return result;
    }
};
